ce/yaml/yaml-types.ts

```diff
@@ -0,0 +1,189 @@
+    this.node = node;
+    if (!(<NodeFactory<ThisType>>(this.constructor)).create) {
+      throw new Error(`class ${this.constructor.name} is missing implementation for create`);
+    }
+  }
+
+  get root(): Yaml {
+    return this.parent ? this.parent.root : this;
+  }
+
+  protected createNode(): ThisType {
+    return (<NodeFactory<ThisType>>this.constructor).create();
+  }
+
+  /**@internal*/ static create() {
+    throw new Error('creator not Not implemented on base class.');
+  }
+
+  private _node: ThisType | undefined;
+
+  get node(): ThisType | undefined {
+    if (this._node) {
+      return this._node;
+    }
+
+    if (this.key && isMap(this.parent?.node)) {
+      this._node = <any>this.parent!.node.get(this.key, true);
+    }
+
+    return this._node;
+  }
+
+  set node(n: ThisType | undefined) {
+    this._node = n;
+  }
+
+  sourcePosition(key?: string | number): Range | undefined {
+    if (!this.node) {
+      return undefined;
+    }
+    if (key !== undefined) {
+      if ((isMap(this.node) || isSeq(this.node))) {
+        const node = <Node>this.node.get(<any>key);
+        if (node) {
+          return node.range || undefined;
+        }
+        return undefined;
+      }
+      if (isScalar(this.node)) {
+        throw new Error('Scalar does not have a key to get a source position');
+      }
+    }
+    return this.node?.range || undefined;
+  }
+
+  /** will dispose of this object if it is empty (or forced) */
+  dispose(force = false, deleteFromParent = true) {
+    if ((this.empty || force) && this.node) {
+      if (deleteFromParent) {
+        this.parent?.deleteChild(this);
+      }
+      this.node = undefined;
+    }
+  }
+
+  /** if this node has any data, this should return false */
+  get empty(): boolean {
+    if (isCollection(this.node)) {
+      return !(this.node?.items.length);
+    }
+    return !isNullish(this.node?.value);
+  }
+  /** @internal */ get exists(): boolean {
+    if (this.node) {
+      return true;
+    }
+    // well, if we're lazy and haven't instantiated it yet, check if it's created.
+    if (this.key && isMap(this.parent?.node)) {
+      this.node = <any>this.parent!.node.get(this.key);
+      if (this.node) {
+        return true;
+      }
+    }
+    return false;
+  }
+  /** @internal */ assert(recreateIfDisposed = false, node = this.node) {
+    if (this.node && this.node === node) {
+      return; // quick and fast
+    }
+
+    if (recreateIfDisposed) {
+      // ensure that this node is the node we're supposed to be.
+      this.node = node;
+
+      if (this.parent) {
+        // ensure that the parent is not disposed
+        this.parent.assert(true);
+
+        if (isMap(this.parent.node)) {
+          if (this.key) {
+            // we have a parent, and the key, we can add the node.
+            // let's just check if there is one first
+            const n = this.parent.node.get(this.key);
+            this.parent!.assert(true);
```

Comment:
Shouldn't need the non-null (!) assertion.

Reply:
fixed